kyo-core/shared/src/test/scala/kyoTest/layersTest.scala

```diff
@@ -0,0 +1,257 @@
+                assert(env.get[Int] == 2)
+                assert(varI == 2)
+            })
+        }
+        "Fibers" in runJVM {
+            val slow = Fibers.delay(50.millis)(IOs("slow"))
+            val fast = IOs("fast")
+            val f    = Layers(Fibers.race(fast, slow))
+
+            f.run.map(env => assert(env.get[String] == "fast"))
+        }
+    }
+    "make" - {
+        "none" in {
+            assertDoesNotCompile("""Layers.make[String]()""")
+        }
+        "missing Target" in {
+            val a = Layers(true)
+            assertDoesNotCompile("""Layers.make(a)""")
+        }
+        "circular dependency" in {
+            val a = Layers.from((s: String) => s.size)
+            val b = Layers.from((i: Int) => i % 2 == 0)
+            val c = Layers.from((b: Boolean) => b.toString)
+
+            assertDoesNotCompile("""Layers.make[String](a, b, c)""")
+        }
+        "missing input" in {
+            val a = Layers.from((s: String) => s.size)
+            val b = Layers.from((i: Int) => i % 2 == 0)
+
+            assertDoesNotCompile("""Layers.make[Boolean](a, b)""")
+        }
+        "missing output" in {
+            val a = Layers(42)
+            val b = Layers.from((i: Int) => i.toDouble)
+            val c = Layers.from((d: Double) => d.toLong)
+
+            assertDoesNotCompile("""Layers.make[String](a, b, c)""")
+        }
+        "complex layer" in run {
+            val s = Layers("Start")
+            val i = Layers.from((s: String) => s.size)
+            val b = Layers.from((i: Int) => i % 2 == 0)
+            case class Several(s: String, i: Int, b: Boolean)
+            val several = Layers.from((s, i, b) => Several(s, i, b))
+
+            trait Parent
+            class Child(val age: Int) extends Parent
+            val child = Layers.from((age: Int) => Child(age * 2))
+
+            case class GrandChild(several: Several, age: Int, parent: Child) extends Parent
+            val grandChild = Layers.from((several, age, child) => GrandChild(several, age, child))
+
+            Layers.make[GrandChild](s, i, b, several, child, grandChild).run.map { env =>
+                val grandchild = env.get[GrandChild]
+                assert(grandchild.several.s == "Start")
+                assert(grandchild.several.i == 5)
+                assert(grandchild.several.b == false)
+                assert(grandchild.age == 5)
+                assert(grandchild.parent.age == 10)
+
+                assert(env.size == 1)
+            }
+        }
+        "effects" in run {
+            class A
+            case class B(a: A)
+            case class C(b: B)
+
+            val a = Layers(IOs(new A))
+            val b = Layers.from((a: A) => Aborts.when(false)("").andThen(B(a)))
+            val c = Layers.from((b: B) => Vars.get[Unit].andThen(C(b)))
+
+            assertCompiles("""Layers.make[C](a, b, c)""")
+        }
+        "pruneable inputs" in pendingUntilFixed {
```

Comment:
Why does this fail? Is it because `TypeMap` is covariant?

Reply:
The goal is that we should fail compilation when too many layers are provided. Currently, we happily only use some of the layers to construct a new layer.